File: Python3/0365_water_and_jug_problem.py
# https://leetcode.cn/problems/water-and-jug-problem/

class Solution:
    def canMeasureWater(self, jug1Capacity, jug2Capacity, targetCapacity):
        """
        Inputs:
            jug1Capacity, jug2Capacity, targetCapacity: int
        Outputs:
            res: bool
        """
        stack = [(0, 0)]
        self.seen = set()
        while stack:
            remain_1, remain_2 = stack.pop()
            if remain_1 == targetCapacity or remain_2 == targetCapacity or remain_1 + remain_2 == targetCapacity:
                return True
            if (remain_1, remain_2) in self.seen:
                continue
            self.seen.add((remain_1, remain_2))
            # Fill 1.
            stack.append((jug1Capacity, remain_2))
            # Fill 2.
            stack.append((remain_1, jug2Capacity))
            # Empty 1.
            stack.append((0, remain_2))
            # Empty 2.
            stack.append((remain_1, 0))
            # Put the water in 1 to 2 until 1 is empty or 2 is full.
            stack.append((remain_1 - min(remain_1, jug2Capacity - remain_2), remain_2 + min(remain_1, jug2Capacity - remain_2)))
            # Put the water in 2 to 1 until 2 is empty or 1 is full.
            stack.append((remain_1 + min(remain_2, jug1Capacity - remain_1), remain_2 - min(remain_2, jug1Capacity - remain_1)))
        return False

File: Cpp/0365_water_and_jug_problem.cpp
// https://leetcode.cn/problems/water-and-jug-problem/

class Solution {
public:
    bool canMeasureWater(int jug1Capacity, int jug2Capacity, int targetCapacity) {
        queue<pair<int,int>> q;
        q.push(make_pair(0,0));
        auto hash_func = [](const pair<int,int> &p) {
            return hash<int>()(p.first) ^ hash<int>()(p.second);
        };
        unordered_set<pair<int,int>, decltype(hash_func)> seen(0, hash_func);
        while (!q.empty()) {
            auto cur = q.front();
            q.pop();
            if (cur.first == targetCapacity || cur.second == targetCapacity || cur.first + cur.second == targetCapacity) {
                return true;
            }
            if (seen.find(cur) != seen.end()) {
                continue;
            }
            seen.insert(cur);
            // Fill 1.
            q.push(make_pair(jug1Capacity, cur.second));
            // Fill 2.
            q.push(make_pair(cur.first, jug2Capacity));
            // Empty 1.
            q.push(make_pair(0, cur.second));
            // Empty 2.
            q.push(make_pair(cur.first, 0));
            // Put the water in 1 to 2 until 1 is empty or 2 is full.
            q.push(make_pair(cur.first-min(cur.first, jug2Capacity-cur.second),
                             cur.second+min(cur.first, jug2Capacity-cur.second)));
            // Put the water in 2 to 1 until 2 is empty or 1 is full.
            q.push(make_pair(cur.first+min(cur.second, jug1Capacity-cur.first),
                             cur.second-min(cur.second, jug1Capacity-cur.first)));
        }
        return false;
    }
};
